examples/pytorch/graphsaint/sampler.py

```diff
@@ -0,0 +1,189 @@
+import math
+import os
+import time
+import torch as th
+import random
+import numpy as np
+import dgl.function as fn
+import dgl
+from dgl.sampling import random_walk, pack_traces
+
+# The base class of sampler
+class SAINTSampler(object):
+    def __init__(self, dn, g, train_nid, node_budget, num_repeat=50):
+        """
+        :param dn: name of dataset
+        :param g: full graph
+        :param train_nid: ids of training nodes
+        :param node_budget: expected number of sampled nodes
+        :param num_repeat: number of repeating sampling one node
+        """
+        self.train_g: dgl.graph = g.subgraph(train_nid)
+        self.dn, self.num_repeat = dn, num_repeat
+        self.node_counter = th.zeros((self.train_g.num_nodes(),))
+        self.edge_counter = th.zeros((self.train_g.num_edges(),))
+        self.prob = None
+
+        graph_fn, norm_fn = self.__generate_fn__()
+
+        if os.path.exists(graph_fn):
+            self.subgraphs = np.load(graph_fn, allow_pickle=True)
```

Comment:
Yes. The original codebase did the same for estimating normalization factors: https://github.com/GraphSAINT/GraphSAINT/blob/6b3465fffcf764b9646bf571e2b4dc5d6aa74f3c/graphsaint/pytorch_version/minibatch.py#L180. It might be possible to add online sampling in the future.

Reply:
If I remember correctly, PyG does something in between: it samples a big bunch of subgraphs for estimating normalization factors, and then samples subgraphs online but assigns the estimated normalization factors to the edges.

Reply:
That's true. The example in PyG README probably does not reproduce the original experiments. Alternatively we can have some examples of GraphSAINT on some OGB datasets for this kind of implementation.